packages/coil-extension/src/background/services/BackgroundScript.ts

```diff
@@ -215,45 +218,50 @@ export class BackgroundScript {
   private setFramesOnChangedListener() {
     // Reset tab state and recheck adapted content when tab changes location
     this.framesService.on('frameChanged', event => {
+      // If not a top level window or its direct descendant, bail early
       if (!(event.frame.top || event.frame.parentFrameId === 0)) {
         return
       }
 
-      const { tabId } = event
-      const status = event.changed.state
+      const { tabId, frameId } = event
+      const changedState = event.changed.state
       const isComplete = event.frame.state === 'complete'
-      const becameComplete = status === 'complete'
+      const becameComplete = changedState === 'complete'
+      const becameLoading = changedState === 'loading'
       const changedUrl = Boolean(event.changed.href)
 
-      // Always get the url from the tab
+      // Always get the current url
       const url = event.frame.href
-      if (status === 'loading' && event.frame.top) {
-        this.setCoilUrlForPopupIfNeeded(tabId, url)
+      if (becameLoading) {
+        if (event.frame.top) {
+          this.setCoilUrlForPopupIfNeeded(tabId, url)
+        }
       }
 
       if (becameComplete || (isComplete && changedUrl)) {
         if (event.frame.top) {
           this.setCoilUrlForPopupIfNeeded(tabId, url)
         }
         const from = `onFrameChanged directly, event=${JSON.stringify(event)}, `
-        const message: CheckAdaptedContent = {
-          command: 'checkAdaptedContent',
-          data: { from }
-        }
-        this.log('sending checkAdaptedContent message', message)
-        this.api.tabs.sendMessage(
-          tabId,
-          message,
-          { frameId: event.frameId },
-          () => {
-            // eslint-disable-next-line @typescript-eslint/no-unused-vars
-            const ignored = this.api.runtime.lastError
-          }
-        )
+        this.checkAdaptedContent(from, tabId, frameId)
       }
     })
   }
 
+  private checkAdaptedContent(from: string, tabId: number, frameId: number) {
+    // Only extension apis can get callbacks for url change events
```

Comment:
aside: should filter this background script side to only check when the url is youtube/twitch